### ISSUE TITLE
Help the "Review each commit separately" workflow

### ISSUE THREAD

**Issue by stefano-maggiolo** [Reviewable/Reviewable#191]
Our workflow uses PRs with potentially several commits, which we usually want to keep separate and review separately (you can think of them as merging "feature branches", even if most of the times it's just a handful of commits). Sometimes we also ask to split or merge commits within a PR. Example: https://reviewable.io/reviews/cms-dev/cms/481#- had one commit, I asked to split it in three.

Now, at the second review, is quite difficult to understand what is going on; in particular, r2, r3, and r4 are never mapped explicitly to the three new commits in the UI (even in the "New commits in r4 [...]" message, the mapping is not clear).

I'm not sure I understand all the ways in which commits reach reviewable, but a few things I'd like to see are the following.
- [ ] Introduce a name for "batch of commits in a PR at a certain time", let's call it batch for the moment.
- [ ] Have the "new commits" message say 
```
new batch with tip <sha> at <time> with 3 commits:
  <sha> <title> <time> mapped to existing rev <revision> because not changed, [...]
  <sha> <title> <time> mapped to <revision>
  <sha> <title> <time> mapped to <revision>
```
- [ ] In the file matrix, there is a visual grouping by batch (might be subtle background color, might be stronger vertical divisions), and a reminder of the mapping between revision and commit.
- [ ] A way of ignoring a batch, or anyway handle the case where the rebase was too heavy and I just want to start review from scratch (in the case above, I cannot review r2 alone against HEAD, as the changes in r1 will sneak in).
- [ ] This is going to be very tricky and I am not sure how to do it or if it is even possible, but a way of reviewing a single commit across batches would be awesome (i.e., if I receive another batch with r5-r7, I'd like to review r5 against r2 as they are the "same" commit (different sha, but same "position" within the batch, or "similar" commit title).

**Comment by pkaminski** [comment]
Hey Stefano, I understand what you're getting at but I'm afraid the fundamental premise of your proposed solution is flawed:  there's no reliable way (that I know of) of detecting batches of rebased commits, or matching them up with their previous incarnations.  The batches that happen to make up pushes are arbitrary, Reviewable doesn't always receive these events from GitHub anyway, and commit timestamps are not always grouped correctly (or even correspond to the order of the commits in the DAG!), so the chances that other commits would end up mixed into the batch are high.  The best idea I've come up with so far is matching on commit title, but that's clearly prone to false positives too.

I agree that I could do a better job of making clear the mapping between revisions and commits.  I'm not clear on how your proposed message format would improve matters, though.  I'm also not sure if you're asking for a change from the current model, which is a sequence of commits that's partitioned into revisions (according to a bunch of heuristics), and where the head of the sequence becomes immutable as revisions are "captured" (exposed to anyone but their author, or otherwise referenced).

> in the case above, I cannot review r2 alone against HEAD, as the changes in r1 will sneak in

This isn't (or shouldn't be) true.  If you select ⊥ and r2 as your diff bounds, you'll get exactly the diff between r2 and its base (which may be different from r1's base), with no traces of r1.  If you're seeing something else then it's a bug.

**Comment by stefano-maggiolo** [comment]
I would be happy with at least some way of seeing the association commit <-> revisions.

More on the issue side, regarding the example of changes from r1 sneaking in when seeing HEAD vs r2: I think you are right in general, but that might be not true when a file was created in r1, see for example https://reviewable.io/reviews/cms-dev/cms/481 for AddTeam.py: this file is created in r1, but is not present in either r2 or r3. When I review HEAD to r3 (which incidentally is the first commit after a rebase), the summary says that AddTeam is in that revision. Is this a real issue?

**Comment by pkaminski** [comment]
Sorry for taking so long in getting back to you.  You're right, diffing r3 against HEAD in your example above will include AddTeam.py.  This is working as intended -- the file is not "not present" in r2 and r3, it's simply unchanged since r1 (but still in the PR).  The file would be hidden if you diffed r3 against r2, as would be normal for a commit-by-commit review.

In diffing r3 against HEAD, was your intention to "start the review fresh" after the r3 rebase?  If so, I'd argue that it's right for the file to be present in the diff, since if you continued a commit-by-commit review from this point on (diffing rN vs rN-1) and the file didn't change again you'd miss out on reviewing it altogether.  On the other hand, if you had already reviewed the file at r1, the diff at r3 vs HEAD will be the same and this will be indicated by a green "mark reviewed" button -- so you'll know you can safely skip it.  Are you perhaps asking for a way to filter out previously reviewed diffs?

In summary: working as intended, but if I can better understand your intentions perhaps I can improve the design.  :smile:

**Comment by stefano-maggiolo** [comment]
Sorry, I missed answering this in a timely fashion. My intention is the first you mention, "start the review fresh".

You argue that it's right for the file to be present in HEAD-r3 because the file is added in r1 and never touched again. But looking back at git, the parent commit of r3 is HEAD, so the tree at r3 doesn't include the file, and therefore shouldn't be included (at least, my interpretation :). Again, I don't know much about the way you get data from github, but I suppose this kind of analysis is difficult from your second comment in this thread.

If what I hope is not possible, looking at the marker "file already reviewed" seems a good idea, thank you.

**Comment by pkaminski** [comment]
Sorry, I think this issue is setting a new record for late replies on my part...

I dug into the git data a bit and, AFAICT, Reviewable's interpretation is correct.  r3 is commit `26ebd3f` with base and parent `a21e5d1`.  The file `cmscontrib/AddTeam.py` is present in `26ebd3f`'s file tree (root at `223ef9e`) with sha `40b5da6` but not present in `a21e5d1`'s file tree (root at `04db74a`).  Hence, as of r3, the file should still be considered as added to the repo.  (The file's also not present in r1's base tree, but I think we agree on that.)

All that said, I agree that while Reviewable deals reasonably well with rebasing and per-commit reviews, combining the two can result in some pretty confusing situations.  I now think that in many common cases matching rebased commits on their commit message would actually work reasonably well, so I'll try to work that in and provide a cue in the UI as to which revisions are "chained" together in this way.  Some tweaks to the commit compaction algorithm are also in order since by default it'll try to collapse rebased commits even in "review each commit separately" mode.

I'll keep this issue open to track that work.

**Comment by pkaminski** [comment]
I decided to just get this done.  Reviewable will now try to detect matching commits across rebases, indicate the correspondence above the file matrix and on hover in individual file revision tables, and automatically select matching commits when diffing.  The fact that a revision is probably a rebase will now also be indicated in the revision's description in the automated top-level comment.

Let me know if this works well for you, or if it's at least a step in the right direction.  :smile:

**Comment by stefano-maggiolo** [comment]
Yay!
